File: packages/core/src/recipe/loader.ts
// Load a recipe BUNDLE from a directory: parse + validate its manifest and read
// the build context (Dockerfile + assets) into memory. Recipes are small. A bundle
// is what an instance is created from (it lives at `<instance>/app/`); the deployed
// unit is an Instance (see instance.ts). Used by ingestion to validate, and by
// instance loading to read the embedded bundle.

import { BRAND } from "../brand.ts";
import { CompositzError } from "../errors.ts";
import type { BuildFile } from "../build.ts";
import { type Manifest, parseManifest } from "./manifest.ts";

export interface Recipe {
  id: string;
  /** Recipe directory, normalized to forward slashes. */
  dir: string;
  manifest: Manifest;
  /** Build-context files with POSIX-relative paths (excludes the manifest). */
  context: BuildFile[];
}

export async function loadRecipe(dir: string): Promise<Recipe> {
  const root = dir.replaceAll("\\", "/").replace(/\/+$/, "");
  const manifestPath = `${root}/${BRAND.manifestFile}`;

  let text: string;
  try {
    text = await Deno.readTextFile(manifestPath);
  } catch {
    throw new CompositzError(`recipe manifest not found: ${manifestPath}`);
  }
  const manifest = parseManifest(text);

  const context: BuildFile[] = [];
  for await (const file of walk(root)) {
    if (file.rel === BRAND.manifestFile) continue; // the manifest is not build context
    if (file.rel.split("/").some((seg) => seg.startsWith("."))) continue; // skip dotfiles/dirs
    context.push({ path: file.rel, data: await Deno.readFile(file.abs) });
  }

  // `image`-based recipes have no build context to validate.
  const dockerfile = manifest.build?.dockerfile;
  if (dockerfile && !context.some((f) => f.path === dockerfile)) {
    throw new CompositzError(
      `recipe "${manifest.id}": Dockerfile "${dockerfile}" not found in ${root}`,
    );
  }
  return { id: manifest.id, dir: root, manifest, context };
}

async function* walk(
  base: string,
  dir: string = base,
): AsyncGenerator<{ abs: string; rel: string }> {
  for await (const entry of Deno.readDir(dir)) {
    const abs = `${dir}/${entry.name}`;
    if (entry.isDirectory) {
      yield* walk(base, abs);
    } else if (entry.isFile) {
      yield { abs, rel: abs.slice(base.length + 1) };
    }
  }
}

File: crates/core/src/recipe/loader.rs
//! Load a recipe BUNDLE from a directory: parse + validate its manifest and read
//! the build context (Dockerfile + assets) into memory. Recipes are small.
//!
//! Ported from `packages/core/src/recipe/loader.ts`. A bundle is what an instance
//! is created from (it lives at `<instance>/app/`); the deployed unit is an
//! [`crate::recipe::instance::Instance`]. Used by ingestion to validate, and by
//! instance loading to read the embedded bundle.

use crate::Error;
use crate::brand;
use crate::build::BuildFile;
use crate::recipe::manifest::{Manifest, parse_manifest};
use crate::recipe::norm_dir;

/// A validated recipe bundle: its manifest and in-memory build context.
#[derive(Debug, Clone)]
pub struct Recipe {
    pub id: String,
    /// Recipe directory, normalized to forward slashes.
    pub dir: String,
    pub manifest: Manifest,
    /// Build-context files with POSIX-relative paths (excludes the manifest).
    pub context: Vec<BuildFile>,
}

/// Load + validate the recipe bundle rooted at `dir`.
pub fn load_recipe(dir: &str) -> Result<Recipe, Error> {
    let root = norm_dir(dir);
    let manifest_path = format!("{root}/{}", brand::MANIFEST_FILE);

    let text = std::fs::read_to_string(&manifest_path)
        .map_err(|_| Error::Recipe(format!("recipe manifest not found: {manifest_path}")))?;
    let manifest = parse_manifest(&text)?;

    let mut context = Vec::new();
    for (abs, rel) in walk(&root)? {
        if rel == brand::MANIFEST_FILE {
            continue; // the manifest is not build context
        }
        if rel.split('/').any(|seg| seg.starts_with('.')) {
            continue; // skip dotfiles/dirs
        }
        context.push(BuildFile {
            path: rel,
            data: std::fs::read(&abs)?,
        });
    }

    // `image`-based recipes have no build context to validate.
    if let Some(build) = &manifest.build
        && !context.iter().any(|f| f.path == build.dockerfile)
    {
        return Err(Error::Recipe(format!(
            "recipe \"{}\": Dockerfile \"{}\" not found in {root}",
            manifest.id, build.dockerfile
        )));
    }

    Ok(Recipe {
        id: manifest.id.clone(),
        dir: root,
        manifest,
        context,
    })
}

/// Recursively collect every file under `root` as `(absolute, root-relative)`
/// forward-slash paths.
fn walk(root: &str) -> Result<Vec<(String, String)>, Error> {
    let mut out = Vec::new();
    walk_into(root, root, &mut out)?;
    Ok(out)
}

fn walk_into(base: &str, dir: &str, out: &mut Vec<(String, String)>) -> Result<(), Error> {
    for entry in std::fs::read_dir(dir)? {
        let entry = entry?;
        let name = entry.file_name();
        let abs = format!("{dir}/{}", name.to_string_lossy());
        let file_type = entry.file_type()?;
        if file_type.is_dir() {
            walk_into(base, &abs, out)?;
        } else if file_type.is_file() {
            // `base` + '/' is one byte past a valid boundary, so the slice is
            // always on a char boundary (the separator is ASCII '/').
            let rel = abs[base.len() + 1..].to_string();
            out.push((abs, rel));
        }
    }
    Ok(())
}
